docs/usage/integration.md

```diff
@@ -0,0 +1,413 @@
+# Integration
+
+TigerBeetle is designed to guard against bugs not only in its [own code](../TIGER_STYLE.md), but
+at the boundaries, in the application code which interfaces with TigerBeetle.
+This is exhibited by the client's API design, which may be [surprising](#retries) when contrasted
+with a more conventional database.
+
+Strict consistency guarantees at the database level simplify application logic and error handling
+farther up the stack.
+
+
+
+## Consistency
+
+TigerBeetle provides strict serializability (serializability + linearizability) to each
+[client session](#client-sessions).
+
+But consistency models are notoriously arcane — what specifically does this mean for applications
+in practice?
```

Comment:
Woohoo! Clarifying arcaneness

Reply:
Serializability is roughly "as if transactions are executed serially, as if nothing else is executing concurrently, so that a transaction gets a view of all the objects in the world, frozen in time, with nothing concurrently modifying them" (this doesn't prevent the database from executing transactions concurrently if they don't overlap, also doesn't guarantee order in which transactions are executed).

Linearizability is roughly "the execution w.r.t. an object happens between operation request and response (and doesn't jump around in time before or after that)".

These two are at the top of the pyramid of all the different types of consistency models so they're common and not arcane. It's beneath this where things start to get hairy and have different names.

Here's a great explanation from Peter Bailis: http://www.bailis.org/blog/linearizability-versus-serializability/

Reply:
Would be great to bring his definitions in with a link back.

Reply:
Imo serializability is still arcane. That is why this section avoids explaining them at a high level it all — it gives specific behaviors that the application developer can expect from TB.

I'm happy to link to that explanation, but I don't want to distract from the list of behaviors.

Reply:
> so they're common and not arcane

Yeah I agree with DJ that these terms and meanings are still not well known. Maybe "arcane" is too harsh. I've read DDIA and read about databases how many times and I still need the refresher each time I look at it.

Reply:
My original draft said "obtuse" instead of "arcane" :wink:.

Reply:
obtuse is also harsh :D 

Instead of:

> But consistency models are notoriously arcane — what specifically does this mean for applications
in practice?

Maybe

> But consistency models are easy to misunderstand or forget about — what specifically does this mean for applications in practice?

"obtuse" and "arcane" can sound like we're putting down the researchers or database developers. Rather, it's just about being gracious to the audience by way of a reminder

Reply:
"forget about" isn't quite right either, but something like that

Reply:
That's great.

Reply:
:thinking: You're right, this sentence needs to be reworked..

"easy to misunderstand" is a little too close to "hard to understand" — which is accurate from my perspective, but is still too close to the obtuse/arcane.

What do you think of:

> But consistency models are easy to misunderstand and hard to apply.
> What specific guarantees does TigerBeetle provide to applications?

I don't love the "unfamiliar systems" part...

Reply:
Somewhere we should just have a nice section on guarantees w.r.t. consistency, and clearly state "strict serializability" and then explain how this is comprised of "serializability" and "linearizability" and what those mean. I like how Kyle Kingsbury always looks for this when he does engagements—that the database has a clear discussion of guarantees.

It is true also that after these big three, things get into the weeds pretty quickly. For example, even in the DistSys community, not everyone agrees even on the taxonomy.

Reply:
Is TigerBeetle's documentation really the place to explain core dist sys concepts though? I agree we should specify what we provide (strict serializability), but if the reader isn't familar with those, it seems like pointing them in the direction of e.g. Bailis' guide is fine.

Reply:
If the original sentence only said "But consistency models can seem arcane" I'd be fine with everything else as is and also with the Bailis link.

Reply:
> Is TigerBeetle's documentation really the place to explain core dist sys concepts though?

We don't have to go deep. A quick intro to all 3 in the right section in the right doc would be worthwhile. Serializability and linearizability together help to explain strict serializability.

Reply:
Perfect. We should just somewhere have a very visible simple guarantees area, where we highlight "strict serializability" as the consistency level, since that's something that should be on "the outside of the box" so that people know the big ideas. i.e. That they're not getting eventual consistency.